File: boj/gold/16437.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int n;
char type[123457];
int cnt[123457];
vector<int> child[123457];

long long dfs(int n) {
    if (child[n].empty()) {
        if (type[n] == 'W')
            return 0;
        else
            return cnt[n];
    }

    long long ans = 0;
    for (int c : child[n])
        ans += dfs(c);

    if (type[n] == 'W')
        return max(ans - cnt[n], 0LL);
    else
        return ans + cnt[n];
}

int main() {
    cin.tie(nullptr);
    cout.tie(nullptr);
    ios_base::sync_with_stdio(false);

    cin >> n;
    for (int i = 2; i <= n; i++) {
        int parent;
        cin >> type[i] >> cnt[i] >> parent;
        child[parent].push_back(i);
    }

    cout << dfs(1);
}

File: boj/gold/16437.py
import sys

sys.setrecursionlimit(10 ** 7)

n = int(sys.stdin.readline())
cnt = [0] * n
child = [[] for _ in range(n)]

for i in range(1, n):
    type, island, parent = sys.stdin.readline().split()
    island = int(island)
    parent = int(parent)

    if type == 'W':
        cnt[i] = -island
    else:
        cnt[i] = island

    child[parent - 1].append(i)

def dfs(x):
    if len(child[x]) == 0:
        return max(0, cnt[x])

    ans = cnt[x]
    for c in child[x]:
        ans += dfs(c)
    return max(0, ans)

print(dfs(0))